### DISCUSSION TITLE
SeqCLI with complex properties

### DISCUSSION THREAD

**Discussion by CrossBound** [datalust/seq-tickets#2195] [Help] [answered]
I am trying to generate logs using the seqcli command and seem to be having some trouble with non-standard or complex properties. I can't figure out how to create nested properties  or numbers that need to be parsed as a string. 

If I just use something like `-p EntityId=003` this gets parsed and turned into something equivalent to `{"EventId": 3}` which is not what I'm looking for. I want the result to be `{"EntityId": "003"}`. 

For nested properties, I've tried things like `Details.SubjectId=3` but this gets parsed into something equivalent to `{"Details.SubjectId": 3}` which is not what I'm after. I'm wanting to be able to create a property that would look like `{"Details": {"SubjectId": "3"}}`

Is it currently possible to do either of these with the seq CLI??

if not, would it be possible to allow json properties to be passed to the log command? For example, I would love to be able to do this
```
seqcli log -m "some message" -p '{"Details": {"SubjectId": "003"}}'
```
Possibly a new parameter like `-jp` or something instead of using the same `-p` parameter.

**Comment by liammclennan** [comment, marked as answer]
Hi @CrossBound 

`seqcli log` can't do what you want. Here is the [relevant code](https://github.com/datalust/seqcli/blob/dev/src/SeqCli/Cli/Features/PropertiesFeature.cs). Bools get parsed as bools, numbers get parsed as numbers and everything else is a string. 

Fortunately, there is a way. You can use `seqcli ingest` to ingest events. Put your events in a newline delimited text file using [CLEF format](http://clef-json.org/).

<img width="1146" alt="image" src="https://github.com/datalust/seq-tickets/assets/271514/a8174dcd-8382-4793-8f10-86e8c0bb0ee6">

and then:

```
seqcli ingest -i data.json --json
```

<img width="1030" alt="image" src="https://github.com/datalust/seq-tickets/assets/271514/bd6c386c-ff59-4f61-aea1-9b3e11e3b059">

**Reply by CrossBound** [reply]
That's disappointing. I would like to suggest a feature request for this as it would be beneficial. I understand you can use ingest but it's a real pain to have to fill in a file with all the extra information (like timestamp, etc.) and then parse it in just to get a message into Seq. For example, that prevents me from having a script or something that generates these and pulls them in without having to go through the trouble of writing out CLEF formatted lines and then reading them back in.

**Reply by liammclennan** [reply]
It doesn't have to be a file. `seqcli` can also read from standard input. Piping JSON through scripts is common and reasonably painless. Otherwise, feel free to submit a PR.

**Reply by nblumhardt** [reply]
We've recently ported some of `seqcli`'s internals to use the shared `Seq.Syntax` package, which has some features we could try using for this.... :thinking: 

In a nutshell, `Seq.Syntax` can construct objects and arrays, has rigorous support for typed values, and also a bunch of handy built-in functions. What if we added an additional parameter to `seqcli log`, let's call it `--properties`, whose value would be a `Seq.Syntax` expression evaluating to an object whose properties would be added to the event?

```
seqcli log -m "some message" --properties "{Details: {SubjectId: '003'}}"
```

The value is specified using the Seq query language syntax, rather than JSON, which plays fairly nicely with shells that accept `"`-quoted strings.

Behind the scenes, all that's required is to construct the `LogEvent` that the `seqcli log` command does today, evaluate the contents of `--properties` over it, and then add the resulting object's members to the `LogEvent` before shipping it off to Seq.

I think the principle of "more-specific overrides less-specific" suggests that any entries in `--properties` that already appear on the event, e.g. via `-p Name=Value`, should just be ignored.

Keen to see this made possible - what do you think?
